core/shared/src/main/scala/cats/parse/Parser.scala

```diff
@@ -921,23 +924,33 @@ object Parser {
       error.input.map(input => (input, error.failedAtOffset, error.expected))
   }
 
-  object Error extends Serializable {
+  object Error
+      extends scala.runtime.AbstractFunction2[Int, NonEmptyList[Expectation], Error]
+      with Serializable {
     def apply(failedAtOffset: Int, expected: NonEmptyList[Expectation]): Error =
       new Error(None, failedAtOffset, expected)
 
     def apply(input: String, failedAtOffset: Int, expected: NonEmptyList[Expectation]): Error =
       new Error(Some(input), failedAtOffset, expected)
 
-    def unapply(error: Error): Some[(Int, NonEmptyList[Expectation])] =
+    def unapply(error: Error): Option[(Int, NonEmptyList[Expectation])] =
       Some((error.failedAtOffset, error.expected))
 
     implicit val catsShowError: Show[Error] =
       new Show[Error] {
         def show(error: Error): String = {
+          val nl = "\n"
+
+          def errorMsg = {
+            val expectations = error.expected.map(e => s"* ${e.show}").toList.mkString(nl)
```

Comment:
nit, `error.expected.toList.iterator.map(...).mkString` will avoid reallocating the whole list.